apollo-federation/src/query_plan/correctness/response_shape_test.rs

```diff
@@ -0,0 +1,518 @@
+        id: ID!
+        data(arg: Int!): String!
+        t: Int!
+    }
+
+    type X implements J {
+        id: ID!
+        data(arg: Int!): String!
+        object(id: ID!): J!
+        x: String!
+    }
+
+    type Y {
+        id: ID!
+        y: String!
+    }
+
+    type Z implements J {
+        id: ID!
+        data(arg: Int!): String!
+        object(id: ID!): J!
+        z: String!
+    }
+
+    union U = R | S | X
+    union V = R | S | Y
+
+    directive @mod(arg: Int!) on FIELD
+"#;
+
+fn response_shape(op_str: &str) -> response_shape::ResponseShape {
+    let schema = Schema::parse_and_validate(SCHEMA_STR, "schema.graphql").unwrap();
+    let schema = ValidFederationSchema::new(schema).unwrap();
+    let op = ExecutableDocument::parse_and_validate(schema.schema(), op_str, "op.graphql").unwrap();
+    response_shape::compute_response_shape_for_operation(&op, &schema).unwrap()
+}
+
+//=================================================================================================
+// Basic response key and alias tests
+
+#[test]
+fn test_aliases() {
+    let op_str = r#"
+        query {
+            test_i {
+                data(arg: 0)
+                alias1: data(arg: 1)
+                alias2: data(arg: 1)
+            }
+        }
+    "#;
+    insta::assert_snapshot!(response_shape(op_str), @r###"
+    {
+      test_i -----> test_i {
+        data -----> data(arg: 0)
+        alias1 -----> data(arg: 1)
+        alias2 -----> data(arg: 1)
+      }
+    }
+    "###);
+}
+
+//=================================================================================================
+// Type condition tests
+
+#[test]
+fn test_type_conditions_over_multiple_different_types() {
+    let op_str = r#"
+        query {
+            test_i {
+                ... on R {
+                    data(arg: 0)
+                }
+                ... on S {
+                    data(arg: 1)
+                }
+            }
+        }
+    "#;
+    insta::assert_snapshot!(response_shape(op_str), @r###"
+    {
+      test_i -----> test_i {
+        data -may-> data(arg: 0) on R
+        data -may-> data(arg: 1) on S
+      }
+    }
+    "###);
+}
+
+#[test]
+fn test_type_conditions_over_multiple_different_interface_types() {
+    // These two intersections are distinct type conditions.
+    // - `U ∧ I` = {R, S}
+    // - `U ∧ J` = `U` = {R, S, X}
+    let op_str = r#"
+        query {
+            test_u {
+                ... on I {
+                    data(arg: 0)
+                }
+                ... on J {
+                    data(arg: 0)
+                }
+            }
+        }
+    "#;
+    // Note: The set {R, S} has no corresponding named type definition in the schema, while
+    //       `U ∧ J` is just the same as `U`.
+    insta::assert_snapshot!(response_shape(op_str), @r###"
+    {
+      test_u -----> test_u {
+        data -may-> data(arg: 0) on I ∧ U = {R, S}
+        data -may-> data(arg: 0) on U
+      }
+    }
```

Comment:
This needs a bit of clarification, or more clearly named types. It becomes confusing to reason about single letter types in sets and intersections 😅 . For example, we can use the `Pet` schema that graphql js uses, or expand on the startstuff schema to add more unions?

Also, what does the `----->` actually denote in these snapshots? I take it `-may->` is an option? So, `test_u` has two possible options?

Reply:
I wrote some design doc ([google doc](https://docs.google.com/document/d/1qlBLEfu4wRyvLvUrrWqR7MGnap5nsTivRao9cZAFVfI/edit?usp=drive_link)).